File: keybinds_searcher.py
#!/usr/bin/env python3
import subprocess
import tempfile
import os

MODIFIER_MAP = {1: "Shift", 4: "Control", 8: "Alt", 64: "Super"}

# You can add your comments here for any keybind combo
COMMENT_OVERRIDES = {
    # workspaces
    "Super + 1": "move to workspace 1 (1 on first monitor)",
    "Super + Shift + 1": "move current window to workspace 1 (1 on first monitor)",
    "Alt + 1": "move to workspace 11 (1 on second monitor)",
    "Alt + Shift + 1": "move current window to workspace 11 ( 1 on second monitor)",
    "Alt + Super + 1": "move to workspace 21 (1 on third monitor)",
    "Alt + Super + Shift + 1": "move current window to workspace 21 ( 1 on third monitor)",
    # Main
    "Super + Q": "Hyprland keybinds searcher help",
    "Super + D": "Application Launcher. Search",
    "Super + W": "Open Internet Browser (Firefox/Zen/LibreWolf)",
    "Super + Return": "Open a terminal emulator. A shell. Cmd. Command",
    "Super + semicolon": "Close focused window. ; Exit, Kill",
    # Other terminals
    "Super + P": "Powermenu, log off, shutdown, reboot",
    "Super + H": "Open a terminal emulator. A shell. Cmd. Command",
    "Super + Less": "Open a terminal (no tmux)",
    "Shift + Super + Less": "Open a different terminal (no tmux)",
    # Add more as needed
    "Super + F1": "Buggy keyboard test",
}


def modmask_to_modifiers(modmask):
    modmask = int(modmask)
    mods = []
    for bit, name in MODIFIER_MAP.items():
        if (modmask & bit) == bit:
            mods.append(name)
    return mods


def parse_hyprctl_binds(text):
    binds = []
    blocks = text.strip().split("\nbind\n")
    for block in blocks:
        lines = block.strip().splitlines()
        data = {}
        for line in lines:
            line = line.strip()
            if not line or ":" not in line:
                continue
            k, v = line.split(":", 1)
            data[k.strip()] = v.strip()

        if not data:
            continue

        mods = modmask_to_modifiers(data.get("modmask", "0"))
        key = data.get("key", "")
        dispatcher = data.get("dispatcher", "")
        arg = data.get("arg", "")

        binds.append([mods + [key], [dispatcher, arg]])
    return binds


def format_bind_line(bind):
    keys = bind[0]
    cmd, arg = bind[1]

    combo = " + ".join(keys)
    base = f"{combo} = {cmd}"
    if arg:
        base += f" {arg}"

    # Check for manual comment
    comment = COMMENT_OVERRIDES.get(combo)
    if comment:
        base += f"    # {comment}"

    return base.strip()


def main():
    try:
        output = subprocess.check_output(["hyprctl", "binds"], text=True)
    except subprocess.CalledProcessError as e:
        print(f"Failed to run hyprctl: {e}")
        return

    binds = parse_hyprctl_binds(output)

    with tempfile.NamedTemporaryFile(mode="w+", delete=False) as tmpfile:
        for bind in binds:
            line = format_bind_line(bind)
            tmpfile.write(line + "\n")

        tmpfile.flush()
        tmpname = tmpfile.name

    try:
        selected = subprocess.check_output(["fzf", "--ansi"], text=True, stdin=open(tmpname))
        print(f"Selected bind:\n{selected.strip()}")
    except subprocess.CalledProcessError:
        print("No selection made or fzf was interrupted.")

    os.unlink(tmpname)


if __name__ == "__main__":
    main()

File: keybinds_searcher.lua
local pickers = require("telescope.pickers")
local finders = require("telescope.finders")
local conf = require("telescope.config").values
local previewers = require("telescope.previewers")

local Job = require("plenary.job")

local function modmask_to_modifiers(modmask)
	local MODIFIER_MAP = {
		[1] = "Shift",
		[4] = "Control",
		[8] = "Alt",
		[64] = "Super",
	}
	local mods = {}
	local mm = tonumber(modmask) or 0
	for bit, name in pairs(MODIFIER_MAP) do
		if (mm & bit) == bit then
			table.insert(mods, name)
		end
	end
	return mods
end

local function parse_binds(output)
	local binds = {}
	for block in output:gmatch("bind\n(.-)\n\n") do
		local modmask = block:match("modmask:%s*(%d+)")
		local key = block:match("key:%s*(%S+)")
		local dispatcher = block:match("dispatcher:%s*(%S+)")
		local arg = block:match("arg:%s*(.-)\n") or ""

		if modmask and key and dispatcher then
			local mods = modmask_to_modifiers(modmask)
			table.insert(mods, key)
			local combo = table.concat(mods, " + ")
			local line = combo .. " = " .. dispatcher .. (arg ~= "" and (" " .. arg) or "")
			table.insert(binds, line)
		end
	end
	return binds
end

local function hyprland_keybinds_picker(opts)
	opts = opts or {}
	Job:new({
		command = "hyprctl",
		args = { "binds" },
		on_exit = function(j, return_val)
			if return_val ~= 0 then
				print("Failed to run hyprctl binds")
				return
			end
			local result = table.concat(j:result(), "\n")
			local items = parse_binds(result)
			pickers
				.new(opts, {
					prompt_title = "Hyprland Keybinds",
					finder = finders.new_table({
						results = items,
					}),
					sorter = conf.generic_sorter(opts),
					previewer = previewers.new_buffer_previewer({
						define_preview = function(self, entry, status)
							-- preview just shows the line for now
							vim.api.nvim_buf_set_lines(self.state.bufnr, 0, -1, false, { entry.value })
						end,
					}),
				})
				:find()
		end,
	}):start()
end

return {
	hyprland_keybinds_picker = hyprland_keybinds_picker,
}
